README.md
# premium-sp-blob


release-notes
Premium Speed Profiles Blob - Release History

Version 1.0.4
 - Updated dependency

Version 1.0.3
 - Updated dependencies
 
Version 1.0.2
 - Added convenience methods to infer the minimum non-negative speed-value supported inside a daily time-bin
 - Simplified code
 - Made FlexSpeedEncoding public to allow for external usage

Version 1.0.1
 - Fixed a bug that caused mean speeds above 127 to be de-coded as negative values

Version 1.0.0
 - Initial version, providing utility classes to encode and decode binary PSP blob data


src/test/java/com/tomtom/traffic/iqr/io/premium/blob/VarIntEncodingTest.java
/*
 * Copyright (C) 2020 TomTom N.V. (www.tomtom.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tomtom.traffic.iqr.io.premium.blob;

import static org.junit.Assert.assertEquals;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

import org.junit.Test;

public class VarIntEncodingTest {

    @Test
    public void encodingOfNegativeValuesYieldsThreeBytes() throws IOException {
        DataOutputStream myOutputStream;
        ByteArrayOutputStream myByteArrayStream;
        for (short value = Short.MIN_VALUE; value < 0; ++value) {
            myByteArrayStream = new ByteArrayOutputStream();
            myOutputStream = new DataOutputStream(myByteArrayStream);
            VarIntEncoding.encode(value, myOutputStream);
            myOutputStream.flush();
            byte[] written = myByteArrayStream.toByteArray();
            assertEquals(
                    String.format("Unexpected number of bytes in VarInt encoding of %d - expected 3, got %d",
                            value, written.length),
                    3, written.length);
        }
    }

    @Test
    public void encodingOfValidValuesYieldsCorrectNumberOfBytes() throws IOException {
        DataOutputStream myOutputStream;
        ByteArrayOutputStream myByteArrayStream;
        for (int intValue = 0; intValue <= Short.MAX_VALUE; ++intValue) {
            myByteArrayStream = new ByteArrayOutputStream();
            myOutputStream = new DataOutputStream(myByteArrayStream);
            short value = (short) intValue;
            VarIntEncoding.encode(value, myOutputStream);
            myOutputStream.flush();
            byte[] written = myByteArrayStream.toByteArray();
            int expectedBytes = expectedNumberOfBytes(value);
            assertEquals(
                    String.format("Unexpected number of bytes in VarInt encoding of %d - expected %d, got %d",
                            value, expectedBytes, written.length),
                    expectedBytes, written.length);
        }
    }

    private static int expectedNumberOfBytes(final short value) {
        if (value <= 0x7f) {
            return 1;
        }
        if (value <= 0x3fff) {
            return 2;
        }
        return 3;
    }

    @Test
    public void decodingEncodedValuesYieldsOriginalValue() throws IOException {
        DataOutputStream myOutputStream;
        ByteArrayOutputStream myByteArrayStream;
        DataInputStream myInputStream;
        ByteArrayInputStream myByteArrayInputStream;

        for (int intValue = Short.MIN_VALUE; intValue <= Short.MAX_VALUE; ++intValue) {
            myByteArrayStream = new ByteArrayOutputStream();
            myOutputStream = new DataOutputStream(myByteArrayStream);
            short value = (short) intValue;
            VarIntEncoding.encode(value, myOutputStream);
            myOutputStream.flush();
            byte[] written = myByteArrayStream.toByteArray();

            myByteArrayInputStream = new ByteArrayInputStream(written);
            myInputStream = new DataInputStream(myByteArrayInputStream);
            short result = VarIntEncoding.decode(myInputStream);
            assertEquals(
                    String.format("Encoding/decoding of %d yields %d", value, result), value, result);
        }
    }
}

src/test/java/com/tomtom/traffic/iqr/io/premium/blob/ZigZagEncodingTest.java
/*
 * Copyright (C) 2020 TomTom N.V. (www.tomtom.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tomtom.traffic.iqr.io.premium.blob;

import static org.junit.Assert.assertEquals;

import org.junit.Test;

public class ZigZagEncodingTest {

    @Test
    public void encodingYieldsExpectedValue() {
        for (int orig = Short.MIN_VALUE; orig <= Short.MAX_VALUE; ++orig) {
            short result = ZigZagEncoding.encode16((short) orig);
            if (orig == 0) {
                assertEquals(String.format("ZigZag.encode(%d) == %d != 0.", orig, result), orig, result);
                continue;
            }
            int bitResult = 0xffff & result;
            int expected;
            if (orig < 0) {
                expected = -orig * 2 - 1;
            } else {
                expected = 2 * orig;
            }
            assertEquals(String.format("ZigZag.encode(%d) == %d != %d.", orig, bitResult, expected), bitResult, expected);
        }
    }

    @Test
    public void decodingYieldsExpectedValue() {
        for (short orig = 0; orig <= (Short.MAX_VALUE - 1) / 2; ++orig) {
            int negativeEncoded = 2 * orig + 1;
            int positiveEncoded = 2 * orig;
            short negativeResult = ZigZagEncoding.decode16((short) negativeEncoded);
            short positiveResult = ZigZagEncoding.decode16((short) positiveEncoded);
            assertEquals(
                    String.format("Decoding of %d yields %d instead of %d.", negativeEncoded, negativeResult, -orig - 1),
                    -orig - 1, negativeResult);
            assertEquals(
                    String.format("Decoding of %d yields %d instead of %d.", positiveEncoded, positiveResult, orig),
                    orig, positiveResult);
        }
    }

    @Test
    public void decodingEncodedYieldsOriginalValue() {
        for (int orig = Short.MIN_VALUE; orig <= Short.MAX_VALUE; ++orig) {
            short result = ZigZagEncoding.decode16(ZigZagEncoding.encode16((short) orig));
            assertEquals(String.format("Encoding/decoding of %d yields %d.", orig, result), orig, result);
        }
    }
}

src/test/java/com/tomtom/traffic/iqr/io/premium/blob/FlexSpeedEncodingTest.java
/*
 * Copyright (C) 2020 TomTom N.V. (www.tomtom.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tomtom.traffic.iqr.io.premium.blob;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.Test;

public class FlexSpeedEncodingTest {

    @Test
    public void testMinimumNonZeroSpeedValues() {
        // ensure that the provided MINIMUM_NONZERO_INPUT_VALUE value is not rounded to 0 when encoded
        assertThat(FlexSpeedEncoding.encode(FlexSpeedEncoding.MINIMUM_NONZERO_INPUT_VALUE))
                .isGreaterThan((short)0);
        // ensure that encoding the provided MINIMUM_NONZERO_INPUT_VALUE value indeed decodes back to the provided
        // MINIMUM_NONZERO_OUTPUT_VALUE
        assertThat(FlexSpeedEncoding.decode(FlexSpeedEncoding.encode(FlexSpeedEncoding.MINIMUM_NONZERO_INPUT_VALUE)))
                .isEqualTo(FlexSpeedEncoding.MINIMUM_NONZERO_OUTPUT_VALUE);
        // ensure that the provided MINIMUM_NONZERO_INPUT_VALUE value is not unnecessarily large
        // (anything smaller is indeed rounded to zero)
        double nextSmallerValue = Math.nextAfter(FlexSpeedEncoding.MINIMUM_NONZERO_INPUT_VALUE, Double.NEGATIVE_INFINITY);
        assertThat(FlexSpeedEncoding.encode(nextSmallerValue)).isEqualTo((short)0);
    }

    @Test
    public void encodingOfNegativeValuesYieldsZero() {
        for (double value = -255; value < 0; value += 0.01) {
            // negative values are always encoded (and thus also decoded) as 0
            assertThat(FlexSpeedEncoding.encode(value)).isEqualTo((short)0);
            assertThat(FlexSpeedEncoding.decode(FlexSpeedEncoding.encode(value))).isEqualTo(0.0);
        }
    }

    @Test
    public void encodingOfTooLargeValuesYieldsMax() {
        double value = 255;
        // just testing some exemplary values above 255
        for (int exponent = 0; exponent < 8; ++exponent) {
            // largest value should be 255 ...
            assertThat(FlexSpeedEncoding.decode(FlexSpeedEncoding.encode(value))).isEqualTo(255);
            // ... which corresponds to 10 bits set
            assertThat(FlexSpeedEncoding.encode(value)).isEqualTo((short)0x3ff);
            value *= 2;
        }
    }

    @Test
    public void encodingValidSpeedsRetainsCorrectResolution() {
        // different value-ranges have different resolution
        // the encoded/decoded values in each range should retain this expected resolution

        // values below 4 have a resolution of 1/64
        assertResolutionInValueRange(0.0, 4.0, 1 / 64.0);

        // above that, values between 2^n-1 and 2^n have resolution 2^-(8 - n)
        double lower = 4.0;
        double upper;
        for (int exponent = 3; exponent <= 7; ++exponent) {
            upper = Math.pow(2, exponent);
            double resolution = Math.pow(2, -(8 - exponent));
            assertResolutionInValueRange(lower, upper, resolution);
            lower = upper;
        }
    }

    private static void assertResolutionInValueRange(final double lower, final double upper, final double resolution) {
        for (double value = lower; value <= upper; value += 0.001) {
            int encoded = FlexSpeedEncoding.encode(value);

            // encoded value must at most have 10 bit
            assertThat(encoded >> 10).isEqualTo(0);

            // encoded value must retain the expected resolution
            double expected = Math.round(value / resolution) * resolution;
            double obtained = FlexSpeedEncoding.decode(FlexSpeedEncoding.encode(value));
            assertThat(obtained)
                    .withFailMessage(
                            String.format("Encoding of %g yields %g instead of %g (resolution = %g)",
                                    value, obtained, expected, resolution))
                    .isEqualTo(expected);
        }
    }

    @Test
    public void encodingOfValidSpeedsYieldsTenBit() {
        for (double value = 0.0; value <= 255; value += 0.001) {
            // encoded value must at most have 10 bit
            assertThat(FlexSpeedEncoding.encode(value) >> 10).isEqualTo(0);
        }
    }
}


src/main/java/com/tomtom/traffic/iqr/io/premium/blob/VarIntEncoding.java
/*
 * Copyright (C) 2020 TomTom N.V. (www.tomtom.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tomtom.traffic.iqr.io.premium.blob;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

/**
 * Utility class providing functionality to read/write {@code short} values in Base-128 encoding from/to streams.
 *
 * For details on the Base-128 encoding see <a href="https://en.wikipedia.org/wiki/Variable-length_quantity">here</a>.
 */
class VarIntEncoding {

    private VarIntEncoding() {
        // Hiding the implicit public constructor
    }

    /**
     * Writes the passed {@code short} value to the provided {@link DataOutputStream} in Base-128 encoding.
     *
     * @param value  the value to be encoded and written
     * @param out  the {@code DataOutputStream} to write the value to
     * @throws IOException  if writing to the passed {@code DataOutputStream} fails.
     */
    public static void encode(final short value, final DataOutputStream out) throws IOException {
        int intValue = Short.toUnsignedInt(value);
        int bits = intValue & 0x7f;
        intValue = intValue >>> 7;
        while (intValue > 0) {
            out.writeByte((byte) (bits | 0x80));
            bits = intValue & 0x7f;
            intValue = intValue >>> 7;
        }
        out.writeByte((byte) bits);
    }

    /**
     * Reads a Base-128-encoded value from the passed {@link DataInputStream} and returns the decoded number as {@code short}.
     *
     * @param in  the {@code DataInputStream} to read from
     * @return the decoded value
     * @throws IOException  if reading from the {@code DataInputStream} fails.
     * @throws IllegalArgumentException  if the read-in data cannot be interpreted a Base-128 value
     */
    public static short decode(final DataInputStream in) throws IOException {
        int result = 0;
        int shift = 0;

        while (shift <= 16) {
            byte bits = in.readByte();
            result = result | ((bits & 0x7f) << shift);
            if ((bits & 0x80) == 0) {
                return (short) result;
            }
            shift += 7;
        }

        throw new IllegalArgumentException("Unexpected number of bytes in VarInt-encoded short.");
    }
}


src/main/java/com/tomtom/traffic/iqr/io/premium/blob/ZigZagEncoding.java
/*
 * Copyright (C) 2020 TomTom N.V. (www.tomtom.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tomtom.traffic.iqr.io.premium.blob;

/**
 * Utility class providing functionality to convert values according to the ZigZag encoding.
 *
 * ZigZag encoding maps signed integers to unsigned integers such that signed numbers with a small absolute value have a
 * small encoded unsigned value too. Effectively, it uses the least significant bit as the sign bit instead of the most
 * significant bit as usual. In other words, negative integer values are mapped to odd unsigned values, while positive
 * values are mapped to even unsigned values.
 *
 * For further details on the ZigZag encoding see
 * <a href="https://en.wikipedia.org/wiki/Variable-length_quantity#Zigzag_encoding">here</a>.
 */
class ZigZagEncoding {

    private ZigZagEncoding() {
        // Hiding the implicit public constructor
    }

    /**
     * Performs ZigZag encoding on the passed value.
     *
     * The returned value is to be interpreted as the unsigned ZigZag-representation of the original value. Note that
     * Java does not natively support unsigned types. Both the in- and output value of the method are thus of a
     * signed type, only the interpretation of the bits is changed. Take this into account accordingly and be careful
     * when comparing or casting the returned values.
     *
     * @param value  the signed value to be encoded
     * @return the unsigned ZigZag-encoded representation of the original value
     */
    public static short encode16(final short value) {
        return (short) ((value << 1) ^ (value >> 15));
    }

    /**
     * Interprets the passed number as an unsigned ZigZag-encoded value and returns the respective original value.
     *
     * @param num  the unsigned ZigZag representation of a signed value
     * @return the represented signed value
     */
    public static short decode16(final short num) {
        int intValue = Short.toUnsignedInt(num);
        return (short) ((intValue >>> 1) ^ (-(intValue & 1)));
    }
}


src/main/java/com/tomtom/traffic/iqr/io/premium/blob/PremiumProfileBlobData.java
/*
 * Copyright (C) 2020 TomTom N.V. (www.tomtom.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tomtom.traffic.iqr.io.premium.blob;

import java.util.Optional;
import java.util.function.IntFunction;

/**
 * Data structure for representing the speed information that is comprised inside a Premium Speed-Profile binary blob.
 *
 * A Premium-Speed-Profile (PSP) binary blob contains the following information about a single directed segment:
 *
 * <ul>
 * <li> The two mean speeds for working-days and weekend-days.</li>
 * <li> Daily 24h-speed-profiles for up to 7 days of the week in a given temporal resolution</li>
 * </ul>
 *
 * A binary blob - and hence the corresponding {@code PremiumProfileBlobData} object - might not contain a daily speed
 * profile for each day of the week or even no daily speed profiles at all. Information about the availability of daily
 * profiles as well as the temporal resolution of all potentially contained profile data can be inferred from the object
 * via a set of corresponding methods (see e.g. {@link #hasDailySpeeds()} and {@link #getTimeResolutionMinutes()}).
 *
 * Mean speed data is always provided and can be accessed via {@link #getWeekDaySpeed()} and {@link #getWeekendSpeed()}.
 *
 * {@code PremiumProfileBlobData} objects can be converted to/from binary blobs as using a
 * {@link PremiumProfileBlobConverter}.
 */
public class PremiumProfileBlobData {
    private final int weekDaySpeed;
    private final int weekendSpeed;
    private final int timeResolutionMinutes;
    private final IntFunction<DaySpeedsAccessor> dataAccessor;

    /**
     * @return the mean speed over all working-days of the week in km/h as an integer value.
     */
    public int getWeekDaySpeed() {
        return weekDaySpeed;
    }

    /**
     * @return the mean speed over all weekend-days in km/h as an integer value.
     */
    public int getWeekendSpeed() {
        return weekendSpeed;
    }

    /**
     * @return the temporal resolution in minutes of the Premium Speed-Profiles in this object.
     * E.g. a value of 30 means that the speeds within each daily profile correspond to the following day-time-intervals:
     * 0:00h-0:30h, 0:30h-1:00h, 1:00h-1:30h, 1:30h-2:00h, etc.
     * The time-periods always cover the entire 24h of the day, starting at 0:00h.
     * All contained daily profiles exhibit the same temporal resolution.
     */
    public int getTimeResolutionMinutes() {
        return timeResolutionMinutes;
    }

    /**
     * @return whether daily Premium Speed-Profiles are available for at least one day (true) or not (false).
     *
     * See also {@link #hasDaySpeeds(int)}.
     */
    public boolean hasDailySpeeds() {
        return dataAccessor != null;
    }

    /**
     * Specifies whether a daily Premium Speed-Profile is available for a given day.
     *
     * @param day  the index of the day for which information about availability of Premium Speed-Profiles is desired.
     *          The allowed index values range from 0 to 6: 0 for Sunday, ..., 6 for Saturday. Note the difference to
     *          {@link java.time.DayOfWeek}.
     * @return whether a daily Premium Speed-Profile is available for the specified day (true) or not (false).
     */
    public boolean hasDaySpeeds(final int day) {
        return dataAccessor != null && isNonEmptyAccessor(dataAccessor.apply(day));
    }

    /**
     * Returns the Premium Speed-Profile data for the specified day (if available).
     *
     * @param day  index value from 0 to 6: 0 for Sunday, ..., 6 for Saturday. Note the difference to
     *          {@link java.time.DayOfWeek}.
     * @return if a Premium Speed-Profile for the day exists, it is returned as a double array, where each value
     * represents the speed for the corresponding time period in km/h. The duration of each time period is given by
     * {@link #getTimeResolutionMinutes()} and the periods always cover the entire 24h of the day.
     * The first value is for 0:00, the second is for (0:00 + {@code getTimeResolutionMinutes()}), etc.
     */
    public Optional<double[]> getDaySpeedsAsArray(final int day) {
        if (dataAccessor != null) {
            DaySpeedsAccessor daySpeedsAccessor = dataAccessor.apply(day);
            double[] result = null;
            if (isNonEmptyAccessor(daySpeedsAccessor)) {
                result = new double[daySpeedsAccessor.getTotalBins()];
                for (int bin = 0; bin < result.length; ++bin) {
                    result[bin] = daySpeedsAccessor.getSpeedByBin(bin);
                }
            }
            return Optional.ofNullable(result);
        }
        return Optional.empty();
    }

    //----------------------------------------------------------------------------------
    //----------------- for writing and other TomTom usage -----------------------------

    /**
     * The {@code DaySpeedsAccessor} interface is used in order to allow the speed-profile data to originate from different sources.
     *
     * A single {@code DaySpeedsAccessor} provides the time-bin-wise speeds of a profile for one specific day of the
     * week.
     *
     * An implementation providing access to data stored in double arrays is available
     * (see {@link #arrayToSpeedsAccessor(double[][])}). Provide additional implementations if data stored in other
     * forms shall be processed.
     */
    public interface DaySpeedsAccessor {

        /**
         * @return the number of speed-profile time bins per day
         */
        int getTotalBins();

        /**
         * Returns the profile-speed in km/h at the time-bin of the provided index.
         *
         * @param bin  index of the time-bin for which the speed shall be returned
         * @return the profile-speed in km/h at the time-bin with the provided index
         */
        double getSpeedByBin(final int bin);
    }

    /**
     * Representation of a 2D-array as day-to-{@link DaySpeedsAccessor}.
     *
     * Implement similarly for other underlying data types.
     *
     * @param dayToSpeeds  array: day to time-bin to speed. The assumed indexing of the days is: 0 for Sunday, ...,
     *          6 for Saturday. (Note the difference to {@link java.time.DayOfWeek}.)
     * @return a thin wrapper around the array
     */
    public static IntFunction<DaySpeedsAccessor> arrayToSpeedsAccessor(final double[][] dayToSpeeds) {
        return day -> {
            double[] daySpeeds = dayToSpeeds[day];
            if (daySpeeds != null && daySpeeds.length > 0) {
                return new PremiumProfileBlobData.DaySpeedsAccessor() {
                    @Override
                    public int getTotalBins() {
                        return daySpeeds.length;
                    }

                    @Override
                    public double getSpeedByBin(final int bin) {
                        return daySpeeds[bin];
                    }
                };
            } else {
                return null;
            }
        };
    }

    /**
     * Constructor for {@code PremiumProfileBlobData} objects providing both mean speeds as well as daily speed profiles.
     *
     * The daily time-bin-wise speed values are provided in form of an {@code IntFunction<DaySpeedsAccessor>}
     * (see {@link DaySpeedsAccessor}) that allows to access speed profiles for single days based on their index. For
     * creating a {@code PremiumProfileBlobData} from arrays for the daily speeds, see
     * {@link #arrayToSpeedsAccessor(double[][])} or use the alternative constructor.
     *
     * @param weekDaySpeed  the mean speed for working-days in km/h as an integer value
     * @param weekendSpeed  the mean speed for weekend-days in km/h as an integer value
     * @param dataAccessor  function that returns daily speeds for specific days of the week given by the corresponding
     *          index in form of a {@code DaySpeedsAccessor}
     * @param timeResolutionMinutes  the width of the time-bins in minutes for which speed values are provided.
     *          Note that a valid {@code PremiumProfileBlobData} is expected to contain as many speed values per day
     *          such that assuming this time-span, the entire 24h of this day are covered.
     */
    public PremiumProfileBlobData(
            int weekDaySpeed,
            int weekendSpeed,
            IntFunction<DaySpeedsAccessor> dataAccessor,
            int timeResolutionMinutes) {
        this.weekDaySpeed = weekDaySpeed;
        this.weekendSpeed = weekendSpeed;
        this.dataAccessor = dataAccessor;
        this.timeResolutionMinutes = timeResolutionMinutes;
    }

    /**
     * Constructor for {@code PremiumProfileBlobData} objects that include only mean speeds but no daily profiles.
     *
     * The resulting object will always return false when calling {@link #hasDailySpeeds()} or {@link #hasDaySpeeds(int)}
     * for any index. {@link #getTimeResolutionMinutes()} will always return 0.
     *
     * @param weekDaySpeed  the mean speed for working-days in km/h as an integer value
     * @param weekendSpeed  the mean speed for weekend-days in km/h as an integer value
     */
    public PremiumProfileBlobData(final int weekDaySpeed, final int weekendSpeed) {
        this(weekDaySpeed, weekendSpeed, (IntFunction<DaySpeedsAccessor>) null, 0);
    }

    /**
     * Convenience Constructor for {@code PremiumProfileBlobData} objects providing both mean speeds as well as daily speed profiles.
     *
     * The daily time-bin-wise speed values are provided in form of a two-dimensional {@code double} array, where the
     * first index specifies the day of the week (compare e.g. {@link #getDaySpeedsAsArray(int)} and the second index
     * specifies the time-bin. Internally uses {@link #arrayToSpeedsAccessor(double[][])} to construct a
     * {@code PremiumProfileBlobData} object.
     *
     * @param weekDaySpeed  the mean speed for working-days in km/h as an integer value
     * @param weekendSpeed  the mean speed for weekend-days in km/h as an integer value
     * @param speedProfileData  two-dimensional {@code double} array providing the speeds per day and time bin in km/h
     * @param timeResolutionMinutes the width of the time-bins in minutes for which speed values are provided. Note
     *          that a valid {@code PremiumProfileBlobData} is expected to contain as many speed values per day such
     *          that assuming this time-span, the entire 24h of this day are covered.
     */
    public PremiumProfileBlobData(
            final int weekDaySpeed,
            final int weekendSpeed,
            final double[][] speedProfileData,
            final int timeResolutionMinutes) {
        this(weekDaySpeed, weekendSpeed, arrayToSpeedsAccessor(speedProfileData), timeResolutionMinutes);
    }

    public Optional<DaySpeedsAccessor> getDaySpeeds(final int day) {
        if (dataAccessor != null) {
            DaySpeedsAccessor daySpeeds = dataAccessor.apply(day);
            return isNonEmptyAccessor(daySpeeds) ? Optional.of(daySpeeds) : Optional.empty();
        }
        return Optional.empty();
    }

    private static boolean isNonEmptyAccessor(final DaySpeedsAccessor daySpeeds) {
        return daySpeeds != null && daySpeeds.getTotalBins() > 0;
    }

}


src/main/java/com/tomtom/traffic/iqr/io/premium/blob/FlexSpeedEncoding.java
/*
 * Copyright (C) 2020 TomTom N.V. (www.tomtom.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tomtom.traffic.iqr.io.premium.blob;

/**
 * Utility class providing functionality for encoding speed values as 10-bit numbers with adaptive relative precision.
 *
 * The values are encoded as 10-bit-floating-point values with 3 bits for the exponent and 7 bits for the mantissa.
 *
 * This encoding can represent 1024 values from 0 to 255. The smallest non-zero value that can be represented is 1/64.
 * The provided precision is not uniform but is different for different value ranges as shown in the following:
 *
 * <table>
 * <tr><th>Exponent    </th> <th>Range          </th> <th>Precision </th></tr>
 * <tr><td>0           </td> <td>0-2            </td> <td> 1/64 </td></tr>
 * <tr><td>1           </td> <td>2-4            </td> <td> 1/64 </td></tr>
 * <tr><td>2           </td> <td>4-8            </td> <td> 1/32 </td></tr>
 * <tr><td>3           </td> <td>8-16           </td> <td> 1/16 </td></tr>
 * <tr><td>4           </td> <td>16-32          </td> <td> 1/8  </td></tr>
 * <tr><td>5           </td> <td>32-64          </td> <td> 1/4  </td></tr>
 * <tr><td>6           </td> <td>64-128         </td> <td> 1/2  </td></tr>
 * <tr><td>7           </td> <td>128-255        </td> <td> 1    </td></tr>
 * <caption>Encoding speed values</caption>
 * </table>
 */

public class FlexSpeedEncoding {

    /** The smallest non-zero input value that will not be rounded to zero when using this encoding. */
    public static final double MINIMUM_NONZERO_INPUT_VALUE = 1 / 128.0;

    /** The smallest non-zero value that can be represented using this encoding. */
    public static final double MINIMUM_NONZERO_OUTPUT_VALUE = 1 / 64.0;

    private FlexSpeedEncoding() {
        // Hiding the implicit public constructor
    }

    /**
     * Encodes a given speed value as a 10-bit floating-point value.
     *
     * Values above 255 and below 0 are capped.
     *
     * @param speed  the speed value to be encoded
     * @return the 10-bit floating-point representation of the value in form of a {@code short}
     */
    public static short encode(final double speed) {
        if (speed <= 0) {
            return 0;
        }
        if (speed <= 2) {
            return (short) Math.round(speed * 64);
        }
        double encodedSpeed = speed;
        if (encodedSpeed >= 255) {
            encodedSpeed = 255;
        }
        // It is safe to calculate the exponent this way since speed is > 2 here
        int exponent = 31 - Integer.numberOfLeadingZeros((int)encodedSpeed);
        // (7 - exponent) is never negative here since encodedSpeed is at most 255
        int mantissa = (int) Math.round(encodedSpeed * (1 << (7 - exponent)) - 128);
        return (short) ((exponent << 7) + mantissa);
    }

    /**
     * Interprets the passed {@code short} value as a 10-bit floating-point-value and returns the decoded value as {@code double}.
     *
     * @param raw  the 10-bit floating-point representation to be decoded
     * @return the decoded value as a {@code double}
     */
    public static double decode(final short raw) {
        if (raw == 0) {
            return 0.0;
        }
        int mantissa = raw & 0x7F;
        int exponent = (raw >> 7) & 7;
        if (exponent == 0) {
            return (double)mantissa / (1 << 6);
        }
        return (mantissa + 128.0) / (1 << (7 - exponent));
    }

}
